src/components/application_manager/src/app_launch/apps_launcher.cc

```diff
@@ -0,0 +1,138 @@
+#include <algorithm>
+#include "application_manager/app_launch/apps_launcher.h"
+#include "utils/make_shared.h"
+#include "utils/timer_task_impl.h"
+#include <iostream>
+
+namespace app_launch {
+struct LauncherGenerator {
+  LauncherGenerator(AppsLauncher& apps_laucnher,
+                    connection_handler::ConnectionHandler& connection_handler,
+                    const uint16_t app_launch_max_retry_attempt,
+                    const uint16_t app_launch_retry_wait_time)
+      : apps_laucnher_(apps_laucnher)
+      , connection_handler_(connection_handler)
+      , app_launch_max_retry_attempt_(app_launch_max_retry_attempt)
+      , app_launch_retry_wait_time_(app_launch_retry_wait_time) {}
+  AppsLauncher::LauncherPtr operator()() {
+    return utils::MakeShared<AppsLauncher::Launcher>(
+        apps_laucnher_,
+        connection_handler_,
+        app_launch_max_retry_attempt_,
+        app_launch_retry_wait_time_);
+  }
+
+  AppsLauncher& apps_laucnher_;
+  connection_handler::ConnectionHandler& connection_handler_;
+  const uint16_t app_launch_max_retry_attempt_;
+  const uint16_t app_launch_retry_wait_time_;
+};
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "AppLaunch")
+AppsLauncher::AppsLauncher(
+    connection_handler::ConnectionHandler& connection_handler,
+    const uint16_t max_number_of_ios_device,
+    const uint16_t app_launch_max_retry_attempt,
+    const uint16_t app_launch_retry_wait_time) {
+  sync_primitives::AutoLock lock(launchers_lock_);
+  free_launchers_.resize(max_number_of_ios_device);
+  std::generate(free_launchers_.begin(),
+                free_launchers_.end(),
+                LauncherGenerator(*this,
+                                  connection_handler,
+                                  app_launch_max_retry_attempt,
+                                  app_launch_retry_wait_time));
+}
+
+void AppsLauncher::StartLaunching(ApplicationDataPtr app_data) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(launchers_lock_);
+  DCHECK_OR_RETURN_VOID(!free_launchers_.empty())
+  const AppLaunchers::iterator it = free_launchers_.begin();
+  LauncherPtr app_launcher = *it;
+  works_launchers_.push_back(app_launcher);
+  free_launchers_.erase(it);
+  app_launcher->PosponedLaunch(app_data);
+}
+
+struct AppLauncherFinder {
+  AppLauncherFinder(const ApplicationDataPtr& app_data) : app_data_(app_data) {}
+  bool operator()(const AppsLauncher::LauncherPtr& launcher) const {
+    return *launcher->app_data_ == *app_data_;
```

Comment:
@LevchenkoS Is there any guarantee that `app_data_` and `launcher` hold valid pointers?